src/dune_engine/action.ml

```diff
@@ -263,10 +263,43 @@ let is_useful_to_distribute = is_useful_to true false
 let is_useful_to_memoize = is_useful_to true true
 
 module Full = struct
-  type nonrec t =
-    { action : t
-    ; env : Env.t
-    ; locks : Path.t list
-    ; can_go_in_shared_cache : bool
-    }
+  module T = struct
+    type nonrec t =
+      { action : t
+      ; env : Env.t
+      ; locks : Path.t list
+      ; can_go_in_shared_cache : bool
+      }
+
+    let empty =
+      { action = Progn []
+      ; env = Env.empty
+      ; locks = []
+      ; can_go_in_shared_cache = true
+      }
+
+    let combine { action; env; locks; can_go_in_shared_cache } x =
+      let action =
+        match (action, x.action) with
```

Comment:
This technically violates monoid laws because associativity doesn't hold, i.e. `Progn [a; Progn [b; c]]` is not the same as `Progn [Progn [a; b]; c]`. Perhaps, it's fixable as follows:

```ocaml
let combine x y = match x, y with
  | Progn [], x 
  | x, Progn [] -> x
  | Progn xs, Progn ys -> Progn (xs @ ys)
  | x, y -> Progn [x; y]
```

Although it's still not obvious that associativity holds.

Btw, it seems nice to add this `combine` to `action.ml` and make that a monoid too.

Reply:
Sounds good, I did this.